keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/plugins/ResourceHandler.kt

```diff
@@ -47,12 +47,16 @@ abstract class ResourceHandler<S : ResourceSpec, R : Any>(
    * Resolve and convert the resource spec into the type that represents the diff-able desired
    * state.
    *
-   * The value returned by this method is used as the basis of the diff (with the result of
-   * [current] in order to decide whether to call [create]/[update]/[upsert].
+   * Returns a pair of values, the first of which is the concrete representation of the resource spec which is used as
+   * the basis of the diff (with the result of [current]) in order to decide whether to call [create]/[update]/[upsert],
+   * the second the resolved resource type, which can be useful to understand what defaults and opinions were applied.
    *
    * @param resource the resource as persisted in the Keel database.
    */
-  suspend fun desired(resource: Resource<S>): R = toResolvedType(applyResolvers(resource))
+  suspend fun desired(resource: Resource<S>): Pair<C, Resource<S>> {
```

Comment:
This function now returns a pair of the concrete type (which is mostly backwards compatible with the previous signature) and the resolved resource, which was previously invisible to callers and can be very useful, as it was in this PR.

Reply:
This has implications on our plugin contract. Not sure it's a problem but we should definitely be pretty conservative in changing this API or requiring too much of the implementor.

Reply:
Important points from Rob, but I do think this is a useful change.

Reply:
Yeah, I realized that, and struggled with the idea of making this change. I had initially made `applyResolvers` public and renamed it to `resolve` to allow clients access to the resolved resource, but then ran into name conflicts with the extension functions defined elsewhere. The two things were also very coupled together for good reason, so I ended up with this approach. But I understand it's a braking change. My statement above that it's "mostly backwards compatible" was bullocks. 😛